## Датасет

In [105]:
import pandas as pd
import numpy as np

data = [
    [1, "Москва", "ЦАО", "Арбатская", 2, 54.0, 32.0, 9.0, 24500000],
    [2, "Москва", "САО", "Аэропорт", 1, 38.5, 19.0, 8.5, 12300000],
    [3, "Москва", "ЮЗАО", "Академическая", 3, 72.0, 45.0, 10.0, 19800000],
    [4, "Москва", "ЗАО", "Кунцевская", 2, 60.0, np.nan, 11.0, 22500000],
    [5, "Москва", "ВАО", "Щелковская", 1, 34.0, 18.0, 6.5, 8750000],

    [6, "Санкт-Петербург", "Центральный", "Чернышевская", 2, 58.0, 35.0, 9.0, 18900000],
    [7, "Санкт-Петербург", "Приморский", "Комендантский пр.", 1, 41.0, 20.0, 10.5, 10900000],
    [8, "Санкт-Петербург", "Московский", "Московская", 3, 81.0, 50.0, 12.0, 20500000],
    [9, "Санкт-Петербург", "Невский", np.nan, 2, 49.0, 29.0, 7.0, 9200000],
    [10, "Санкт-Петербург", "Василеостровский", "Василеостровская", 4, 105.0, 68.0, 14.0, 31800000],

    [11, "Казань", "Вахитовский", "Площадь Тукая", 2, 56.0, 33.0, 10.0, 11200000],
    [12, "Казань", "Советский", "Горки", 1, 36.0, 18.5, 8.0, 5700000],
    [13, "Казань", "Ново-Савиновский", "Козья слобода", 3, 78.0, 48.0, 11.5, 13900000],
    [14, "Казань", "Приволжский", np.nan, 2, 52.0, 30.0, np.nan, 7800000],
    [15, "Казань", "Московский", "Яшьлек", 1, 32.0, 17.0, 6.0, 4300000],

    [16, "Екатеринбург", "Центр", "Площадь 1905 года", 2, 61.0, 36.0, 10.0, 11500000],
    [17, "Екатеринбург", "ВИЗ", "Площадь 1905 года", 1, 39.0, 19.0, 9.0, 7600000],
    [18, "Екатеринбург", "Уралмаш", "Уралмаш", 3, 66.0, 42.0, 8.0, 6900000],
    [19, "Екатеринбург", "Академический", np.nan, 2, 55.0, 31.0, 12.0, 8200000],
    [20, "Екатеринбург", "Ботаника", "Ботаническая", np.nan, 74.0, 46.0, 11.0, 10500000],

    [21, "Новосибирск", "Центральный", "Площадь Ленина", 2, 57.0, 34.0, 9.0, 9100000],
    [22, "Новосибирск", "Заельцовский", "Заельцовская", 1, 35.5, 18.0, 7.5, 4700000],
    [23, "Новосибирск", "Октябрьский", "Октябрьская", 3, 70.0, 43.0, 9.5, 10300000],
    [24, "Новосибирск", "Ленинский", np.nan, 2, np.nan, 30.0, 8.0, 6100000],
    [25, "Новосибирск", "Кировский", "Студенческая", 1, 31.0, 16.0, 6.0, 3500000],

    [26, "Нижний Новгород", "Нижегородский", "Горьковская", 2, 59.0, 35.0, 9.0, 8900000],
    [27, "Нижний Новгород", "Советский", "Горьковская", 1, 37.0, 18.0, 8.5, 5100000],
    [28, "Нижний Новгород", "Канавинский", "Московская", 3, 73.0, 45.0, 10.0, 8200000],
    [29, "Нижний Новгород", "Автозаводский", np.nan, 2, 50.0, 29.0, 7.5, 4600000],
    [30, "Нижний Новгород", "Сормовский", "Буревестник", np.nan, 44.0, 25.0, np.nan, 3900000],

    [31, "Краснодар", "Центральный", np.nan, 2, 62.0, 34.0, 13.0, 9500000],
    [32, "Краснодар", "Фестивальный", np.nan, 1, 42.0, 20.0, 12.0, 6100000],
    [33, "Краснодар", "Юбилейный", np.nan, 3, 84.0, 50.0, 14.0, 11200000],
    [34, "Краснодар", "Гидростроителей", np.nan, 2, 54.0, np.nan, 10.5, 5900000],
    [35, "Краснодар", "Пашковский", np.nan, 1, 36.0, 18.0, 9.0, 3700000],

    [36, "Ростов-на-Дону", "Кировский", np.nan, 2, 60.0, 34.0, 12.0, 8200000],
    [37, "Ростов-на-Дону", "Ленинский", np.nan, 1, 40.0, 19.0, 10.0, 5400000],
    [38, "Ростов-на-Дону", "Советский", np.nan, 3, 76.0, 47.0, 11.0, 6500000],
    [39, "Ростов-на-Дону", "Первомайский", np.nan, 2, 51.0, 30.0, 8.0, 4100000],
    [40, "Ростов-на-Дону", "Ворошиловский", np.nan, np.nan, 69.0, 40.0, 13.0, 7900000],

    [41, "Самара", "Ленинский", "Алабинская", 2, 55.0, 32.0, 9.0, 7200000],
    [42, "Самара", "Октябрьский", "Российская", 1, 34.0, 17.0, 7.0, 3900000],
    [43, "Самара", "Промышленный", "Безымянка", 3, 68.0, 43.0, 8.5, 5600000],
    [44, "Самара", "Кировский", np.nan, 2, 48.0, 28.0, np.nan, 3200000],
    [45, "Самара", "Красноглинский", np.nan, 1, np.nan, 18.0, 8.0, 3400000],

    [46, "Пермь", "Ленинский", np.nan, 2, 53.0, 31.0, 9.0, 6100000],
    [47, "Пермь", "Свердловский", np.nan, 1, 33.0, 17.0, 7.0, 3100000],
    [48, "Пермь", "Мотовилихинский", np.nan, 3, 71.0, 44.0, 10.0, 7200000],
    [49, "Пермь", "Индустриальный", np.nan, 2, 46.0, 27.0, 6.5, 2900000],
    [50, "Пермь", "Дзержинский", np.nan, np.nan, 64.0, 38.0, 11.0, 6800000],
]

columns = [
    "id",
    "city",
    "district",
    "metro",
    "rooms",
    "total_area_m2",
    "living_area_m2",
    "kitchen_area_m2",
    "price_rub"
]

df = pd.DataFrame(data, columns=columns)



---



## Задача 1. Посмотреть таблицу

Выведите первые 5 строк таблицы.

**Сделайте вывод:**

- Какие данные хранятся в таблице?
- Какие столбцы есть в датасете?

In [106]:
df.head()

,id,city,district,metro,rooms,total_area_m2,living_area_m2,kitchen_area_m2,price_rub
0,1,Москва,ЦАО,Арбатская,2.0,54.0,32.0,9.0,24500000
1,2,Москва,САО,Аэропорт,1.0,38.5,19.0,8.5,12300000
2,3,Москва,ЮЗАО,Академическая,3.0,72.0,45.0,10.0,19800000
3,4,Москва,ЗАО,Кунцевская,2.0,60.0,NaN,11.0,22500000
4,5,Москва,ВАО,Щелковская,1.0,34.0,18.0,6.5,8750000


###Вывод
В таблице хранятся данные о продажах квартир. В ней представлены следующие столбцы:
- id квартиры
- город
- район
- ближайшая станция метро
- количество комнат
- общая площадь
- жилая площадь
- площадь кухни
- цена квартиры



---



## Задача 2. Размер таблицы

Узнайте, сколько в таблице строк и столбцов.


**Сделайте вывод:**

- Сколько всего квартир в датасете?
- Сколько характеристик описывает каждую квартиру?

In [107]:
df.shape

(50, 9)

In [108]:
rows, cols = df.shape
print(f'В датасете всего {rows} квартир. Каждая из них описывается {cols} характеристиками')

В датасете всего 50 квартир. Каждая из них описывается 9 характеристиками


###Вывод
В датасете всего 50 квартир, каждая из них описывается 9 характеристиками



---



## Задача 3. Пропуски в данных

Посчитайте, сколько пропусков есть в каждом столбце.


**Сделайте вывод:**

- В каких столбцах есть пропуски?
- В каком столбце пропусков больше всего?
- Могут ли пропуски помешать анализу?

In [109]:
df.isna().sum()

,0
id,0
city,0
district,0
metro,22
rooms,4
total_area_m2,2
living_area_m2,2
kitchen_area_m2,3
price_rub,0


###Вывод
Пропуски есть в следующих столбцах:
- ближайшая станция метро
- количество комнат
- общая площадь
- жилая площадь
- площадь кухни

Больше всего пропущенных значений в столбце "Ближайшая станция метро". Это может быть связано с тем, что в городе отсутствует метро. Это наврядли сможет помешать анализу.

Пропуски в столбцах "Количество комнат", "Общая площадь", "Жилая площадь", "Площадь кухни" могут помешать анализу, т.к. эти признаки являются одними из самых главных в формировании цены на квартиру




---



## Задача 4. Квартиры в одном городе

Выведите все квартиры, которые находятся в Москве.


**Сделайте вывод:**

- Сколько квартир продаётся в Москве?
- Какие районы Москвы есть в таблице?
- Какая самая дорогая квартира среди московских?

In [110]:
df_moscow = df[df['city'] == 'Москва']
df_moscow

,id,city,district,metro,rooms,total_area_m2,living_area_m2,kitchen_area_m2,price_rub
0,1,Москва,ЦАО,Арбатская,2.0,54.0,32.0,9.0,24500000
1,2,Москва,САО,Аэропорт,1.0,38.5,19.0,8.5,12300000
2,3,Москва,ЮЗАО,Академическая,3.0,72.0,45.0,10.0,19800000
3,4,Москва,ЗАО,Кунцевская,2.0,60.0,NaN,11.0,22500000
4,5,Москва,ВАО,Щелковская,1.0,34.0,18.0,6.5,8750000


In [111]:
# Выведем количество квартир в Москве
print(f'В Москве всего продается {df_moscow.shape[0]} квартир')

# Найдем уникальные районы Москвы
districts = df_moscow['district'].unique().tolist()
print(f'В таблице есть следующие районы Москвы: {', '.join(districts)}')

В Москве всего продается 5 квартир
В таблице есть следующие районы Москвы: ЦАО, САО, ЮЗАО, ЗАО, ВАО


Самая дорогая квартира:

In [112]:
df_moscow.sort_values('price_rub', ascending=False).head(1)


,id,city,district,metro,rooms,total_area_m2,living_area_m2,kitchen_area_m2,price_rub
0,1,Москва,ЦАО,Арбатская,2.0,54.0,32.0,9.0,24500000


###Вывод
В Москве продается 5 квартир. Они представлены в районах: ЦАО, САО, ЮЗАО, ЗАО, ВАО. Самой дорогой является квартира возле станции метро Арбатская в ЦАО с двумя комнатами и общей площадью 54 м² за 24 500 000 рублей



---



## Задача 5. Недорогие квартиры

Найдите квартиры дешевле 5 000 000 рублей.

Выведите столбцы:

```python
city, district, rooms, total_area_m2, price_rub
```

**Сделайте вывод:**

- В каких городах есть квартиры дешевле 5 млн рублей?
- Сколько таких квартир?
- Есть ли среди них квартиры больше 50 м²?

In [113]:
df_low_price = df.loc[df['price_rub'] < 5000000, ['city', 'district', 'rooms', 'total_area_m2', 'price_rub']]
df_low_price

,city,district,rooms,total_area_m2,price_rub
14,Казань,Московский,1.0,32.0,4300000
21,Новосибирск,Заельцовский,1.0,35.5,4700000
24,Новосибирск,Кировский,1.0,31.0,3500000
28,Нижний Новгород,Автозаводский,2.0,50.0,4600000
29,Нижний Новгород,Сормовский,NaN,44.0,3900000
34,Краснодар,Пашковский,1.0,36.0,3700000
38,Ростов-на-Дону,Первомайский,2.0,51.0,4100000
41,Самара,Октябрьский,1.0,34.0,3900000
43,Самара,Кировский,2.0,48.0,3200000
44,Самара,Красноглинский,1.0,NaN,3400000


In [114]:
# Найдем в каких городах есть квартиры дешевле 5 млн рублей
cities = df_low_price['city'].unique().tolist()
print(f'В этих городах квартиры дешевле 5 млн рублей: {', '.join(cities)}')

# Найдем сколько таких квартир
print(f'Квартир дешевле 5 млн рублей в этом датасете всего {df_low_price.shape[0]} штук')

В этих городах квартиры дешевле 5 млн рублей: Казань, Новосибирск, Нижний Новгород, Краснодар, Ростов-на-Дону, Самара, Пермь
Квартир дешевле 5 млн рублей в этом датасете всего 12 штук


Квартиры больше 50 м²:

In [115]:
df_low_price[df_low_price['total_area_m2'] > 50]

,city,district,rooms,total_area_m2,price_rub
38,Ростов-на-Дону,Первомайский,2.0,51.0,4100000


###Вывод
В датасете представлено 12 квартир стоимостью менее 5 млн рублей. Эти квартиры расположены в городах: Казань, Новосибирск, Нижний Новгород, Краснодар, Ростов-на-Дону, Самара и Пермь.

Даже в этом ценовом диапазоне встречаются довольно просторные квартиры: 51 м² в Ростове‑на‑Дону за 4 100 000 руб.



---



## Задача 6. Сортировка по цене

Отсортируйте квартиры по цене от самой дорогой к самой дешёвой.

Выведите первые 10 строк.

**Сделайте вывод:**

- В каком городе находится самая дорогая квартира?
- Сколько комнат у самой дорогой квартиры?
- Есть ли среди самых дорогих квартир квартиры не из Москвы?

In [116]:
df_most_expencive_apartments = df.sort_values('price_rub', ascending=False).head(10)
df_most_expencive_apartments

,id,city,district,metro,rooms,total_area_m2,living_area_m2,kitchen_area_m2,price_rub
9,10,Санкт-Петербург,Василеостровский,Василеостровская,4.0,105.0,68.0,14.0,31800000
0,1,Москва,ЦАО,Арбатская,2.0,54.0,32.0,9.0,24500000
3,4,Москва,ЗАО,Кунцевская,2.0,60.0,NaN,11.0,22500000
7,8,Санкт-Петербург,Московский,Московская,3.0,81.0,50.0,12.0,20500000
2,3,Москва,ЮЗАО,Академическая,3.0,72.0,45.0,10.0,19800000
5,6,Санкт-Петербург,Центральный,Чернышевская,2.0,58.0,35.0,9.0,18900000
12,13,Казань,Ново-Савиновский,Козья слобода,3.0,78.0,48.0,11.5,13900000
1,2,Москва,САО,Аэропорт,1.0,38.5,19.0,8.5,12300000
15,16,Екатеринбург,Центр,Площадь 1905 года,2.0,61.0,36.0,10.0,11500000
10,11,Казань,Вахитовский,Площадь Тукая,2.0,56.0,33.0,10.0,11200000


In [117]:
# Найдем город, в котором самая дорогая квартира
print('Самая дорогая квартира находится в городе', df_most_expencive_apartments.iloc[0] ['city'])

# Найдем количество комнат в этой квартире
print(f'В самой дорогой квартире {int(df_most_expencive_apartments.iloc[0]['rooms'])} комнаты')

# Найдем в каких еще городах, кроме Москвы, есть самые дорогие квартиры
unique_cities = df_most_expencive_apartments.loc[df_most_expencive_apartments['city'] != 'Москва', 'city'].unique().tolist()
print('Кроме Москвы в список самых дорогих квартир, вошли квартиры в городах', ', '.join(unique_cities))

Самая дорогая квартира находится в городе Санкт-Петербург
В самой дорогой квартире 4 комнаты
Кроме Москвы в список самых дорогих квартир, вошли квартиры в городах Санкт-Петербург, Казань, Екатеринбург


###Вывод
Вывели ТОП-10 самых дорогих квартир. Оказалось, что самой дорогой является квартира из Санкт-Петербурга возле станции метро Василеостровская общей площадью 105 м² за 31 800 000 рублей. В ней 4 комнаты.

Кроме Москвы и Санкт-Петербурга в список с самыми дорогими квартирами, вошли также квартиры, расположенные в Казани и Екатеринбурге



---



## Задача 7. Средняя цена по городам

Посчитайте среднюю цену квартиры для каждого города.

**Сделайте вывод:**

- В каком городе самая высокая средняя цена?
- В каком городе самая низкая средняя цена?
- Насколько сильно отличаются цены между городами?

In [118]:
df.groupby('city')['price_rub'].mean().reset_index().sort_values('price_rub', ascending=False)


,city,price_rub
9,Санкт-Петербург,18260000.0
3,Москва,17570000.0
0,Екатеринбург,8940000.0
1,Казань,8580000.0
2,Краснодар,7280000.0
5,Новосибирск,6740000.0
7,Ростов-на-Дону,6420000.0
4,Нижний Новгород,6140000.0
6,Пермь,5220000.0
8,Самара,4660000.0


In [119]:
# Создадим датасет средней цены по городам
df_mean = df.groupby('city')['price_rub'].mean()

# Найдем в каком городе самая высокая средняя цена
print('Самая высокая средняя цена в городе:', df_mean.idxmax())

# Найдем в каком городе самая низкая средняя цена
print('Самая низкая средняя цена в городе:', df_mean.idxmin())

# Найдем разницу в средней цене между этими городами
price_range = df_mean.max() - df_mean.min()
print('Разница в средней цене на квартиры между этими городами составляет:', price_range)

Самая высокая средняя цена в городе: Санкт-Петербург
Самая низкая средняя цена в городе: Самара
Разница в средней цене на квартиры между этими городами составляет: 13600000.0


###Вывод
Посчитали среднюю цену квартиры для каждого города. Оказалось, что самая высокая средняя цена на квартиру не в Москве, а в Санкт-Петербурге и составляет 18 260 000 рублей. Самая низкая средняя цена в Самаре, и она равна 4 660 000 рублей. Разница между ними составляет 13 600 000 рублей. Это может указывать на то, что в нашем датасете представлены только очень дорогие квартиры из Санкт-Петербурга. Для более объективного анализа нам необходимо иметь более сбалансированную выборку по количеству и ценам на квартиры в каждом городе